pennylane/gradients/gradient_transform.py

```diff
@@ -114,70 +114,94 @@ def assert_no_tape_batching(tape, transform_name):
         )
 
 
-def _gradient_analysis(tape, use_graph=True, grad_fn=None):
-    """Update the parameter information dictionary of the tape with
-    gradient information of each parameter.
+def choose_trainable_params(tape, argnum=None):
+    """Returns a list of trainable parameter indices in the tape.
+
+    Chooses the subset of trainable parameters to compute the Jacobian for. The function
+    returns a list of indices with respect to the list of trainable parameters. If argnum
+    is not provided, all trainable parameters are considered.
+
+    Args:
+        tape (`~.QuantumScript`): the tape to analyze
+        argnum (int, list(int), None): Indices for trainable parameters(s)
+            to compute the Jacobian for.
+
+    Returns:
+        list: list of the trainable parameter indices
+
+    """
+
+    if argnum is None:
+        return [idx for idx, _ in enumerate(tape.trainable_params)]
+
+    if isinstance(argnum, int):
+        argnum = [argnum]
+
+    if len(argnum) == 0:
+        warnings.warn(
+            "No trainable parameters were specified for computing the Jacobian.",
+            UserWarning,
+        )
+
+    return argnum
+
+
+def _try_zero_grad_from_graph_or_get_grad_method(tape, param_index, use_graph=True):
+    """Gets the gradient method of a parameter. If use_graph=True, analyze the
+    circuit graph to find if the parameter has zero gradient.
+
+    Args:
+        tape (`~.QuantumScript`): the tape to analyze
+        param_index (int): the index of the parameter to analyze
+        use_graph (bool): whether to use the circuit graph to find if
+            a parameter has zero gradient
+
     """
+
     # pylint:disable=protected-access
-    if grad_fn is not None:
-        if getattr(tape, "_gradient_fn", None) is grad_fn:
-            # gradient analysis has already been performed on this tape
-            return
+    par_info = tape._par_info[param_index]
 
-        tape._gradient_fn = grad_fn
+    if use_graph:

```

Comment:
Do you know where is it still used? This looks like some legacy behavior with the graph. Maybe we could remove it.

Reply:
I think the idea is that computing the graph is expensive, so even though this is not used anywhere right now, we want to keep this so that at some point maybe we want to give the option of skipping graph analysis. @albi3ro is that accurate?

Reply:
That's the idea, but I have no idea how a developer/ user would turn it off for a workflow. So it may indeed by a dead logical branch.

Reply:
We can pass it as an optional argument to all the gradient methods?

Reply:
Tom and I have discussed this in the past; the reason we always kept this feature in is that we would sometimes find/encounter workflows that had redundancy in the quantum circuit wrt gradients, and so it was worth slightly more classical overhead to reduce the quantum overhead.

If it is not accessible by anyone though, that's another question

Reply:
Let's just leave this as is for now, but leave a note to come back to this later.

Once we can just `null.qubit` in full workflows, I'd like to do some analysis of the classical overheads associated with gradients in large workflows. We've already got a datapoint from AWS that it can be rather memory heavy. Once we do that analysis, we will have the proper information to decide what to do.